### ISSUE TITLE
Enhancement: Support for FILE_SUPPORTS_POSIX_UNLINK_RENAME

### ISSUE THREAD

**Issue by billziss-gh** [billziss-gh/winfsp#397]
As per title.

**Comment by billziss-gh** [comment]
Preliminary notes for implementation. These notes may be incomplete, inaccurate or plain old wrong. May be updated as implementation progresses.

----

## Header changes

- `fsctl.h`:
    - `FSP_FSCTL_VOLUME_PARAMS_V0_FIELD_DEFN`:
        - Add `SupportsPosixUnlinkRename:1`
        - Add `CleanupResultInfo:1`
        - Add `PostCleanupOnDeleteOnly:1`. Can we implement this?
        - Change `KmReservedFlags:3` to `KmReservedFlags:1` or delete (depending on `PostCleanupOnDeleteOnly` feasibility).
        - Likely we should unify `SupportsPosixUnlinkRename` and `CleanupResultInfo`.
    - `FSP_FSCTL_TRANSACT_REQ`:
        - `SetInformation.Delete`: add union to support flags

- `winfsp.h`: Add `FSP_FILE_SYSTEM_INTERFACE::Delete`.
    - `Delete` supersedes `CanDelete` and `SetDelete` when available.
    - File system implementation should look like:
        ```C
        NTSTATUS Delete(FSP_FILE_SYSTEM *FileSystem,
            PVOID FileContext, PWSTR FileName, ULONG Flags)
        {
            switch (Flags)
            {
            case FILE_DISPOSITION_DO_NOT_DELETE:
                // set file disposition flag: do not delete file at Cleanup
                return STATUS_SUCCESS;
            case FILE_DISPOSITION_DELETE:
                // set file disposition flag: delete file at Cleanup
                return STATUS_SUCCESS;
            case FILE_DISPOSITION_DELETE | FILE_DISPOSITION_POSIX_SEMANTICS:
                // delete file now; open handles to file remain valid
                return STATUS_SUCCESS;
            case (ULONG)-1:
                // delete file now; called during Cleanup
                return STATUS_SUCCESS;
            default:
                return STATUS_INVALID_PARAMETER;
            }
        }
        ```

## FSD changes

- `FspFsvolQueryFsAttributeInformation`: add `FILE_SUPPORTS_POSIX_UNLINK_RENAME`
- `FspFsvolSetInformation`: add `FileDispositionInformationEx`.
- `FspFsvolSetInformationComplete`: add `FileDispositionInformationEx`.
- `FspFsvolSetDispositionInformation`: modified to support `FileDispositionInformation` and `FileDispositionInformationEx`.
    - `FileDispositionInformationEx`: should only pass along `FILE_DISPOSITION_DO_NOT_DELETE`, `FILE_DISPOSITION_DELETE`, `FILE_DISPOSITION_POSIX_SEMANTICS`, `FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE`. Other flags should be handled internaly by FSD.
        - `FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE`: should be pruned if `FileDesc->GrantedAccess` does not have `FILE_WRITE_ATTRIBUTES`.
        - `FILE_DISPOSITION_FORCE_IMAGE_SECTION_CHECK`: should perform the `MmFlushImageSection` check.
        - `FILE_DISPOSITION_ON_CLOSE`: ?
- `FspFsvolSetDispositionInformationSuccess`: how to handle additional `FILE_DISPOSITION_*` cases here?
    - When doing POSIX delete, we should post all delete notifications that `FspFsvolCleanupComplete` posts when deleting.
    - When doing POSIX delete, we should remove the file from the Context table.
- `FspFileNodeRename`: what changes are required for rename here?
- `FspFsvolCleanup`: determine if possible to implement the `PostCleanupOnDeleteOnly` optimization flag.
    - Main obstacle is the metadata notifications done in `FspFsvolCleanupComplete`.
- `FspFsvolCleanupComplete`:
    - Support `FILE_CLEANUP_*` flags when `CleanupResultInfo`.
        - Alternatively we can have WinFsp always claim that `FILE_RETURNS_CLEANUP_RESULT_INFO`.
        - When doing POSIX delete, do not post delete notifications.
- `FspFileNodeCleanupComplete`:
        - When doing POSIX delete, do no remove the file from the Context table as it was already removed by `FspFsvolSetDispositionInformationSuccess`.
- `FspIopCompleteIrpEx`: allow `Irp->IoStatus.Information != 0` for `IRP_MJ_CLEANUP`.

## DLL changes

- `FspFileSystemOpSetInformation`: modified to support `FileDispositionInformation` and `FileDispositionInformationEx`.
    - `FileDispositionInformationEx`: `FILE_ATTRIBUTE_READONLY` check should be skipped if `FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE` is set. The `FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE` flag should be pruned.
    - `Delete`: `Delete` supersedes `CanDelete` and `SetDelete` if available.
- `FspFileSystemOpCleanup`: modified to call `Delete(-1)` on `Request->Req.Cleanup.Delete`.
    - `Delete` result should be converted to `FILE_CLEANUP_*` codes and passed to `Response->IoStatus.Information`.

## FUSE

- `SupportsPosixUnlinkRename`, `CleanupResultInfo`, `PostCleanupOnDeleteOnly` should be enabled in the FUSE layer.
- `Delete` call added that does `unlink` or `rmdir` accordingly.
- `Cleanup` can be removed.

## .NET

- Support to add flags such as `SupportsPosixUnlinkRename`, etc.
- `Delete` should be added.

**Comment by billziss-gh** [comment]
This functionality has been implemented and is available in [WinFsp 2021.1 Beta3](https://github.com/billziss-gh/winfsp/releases/tag/v1.10B3).
